### ISSUE TITLE
Building justbuild against fmt 9.1.x fails due to missing fallback formatter

### ISSUE THREAD

**Issue by mhthies** [just-buildsystem/justbuild#11]
I'm trying to package *justbuild* for ArchLinux, linking against native shared libraries, provided by ArchLinux packages, as far as possible. In doing so, I came across yet another problem when building against *fmt* version 9.1.0 (in addition to #10):

```txt
Running ['/usr/bin/clang++', '-std=c++20', '-DBOOTSTRAP_BUILD_TOOL', '-I', '/tmp/tmp1i8ffxxt/src', '-I', '/tmp/tmp1i8ffxxt/dep_includes', '-I', '/usr/include', '-c', '/tmp/tmp1i8ffxxt/src/src/buildtool/main/analyse.cpp', '-o', '/tmp/tmp1i8ffxxt/src/src/buildtool/main/analyse.o'] in '/tmp/tmp1i8ffxxt/src'
In file included from /tmp/tmp1i8ffxxt/src/src/buildtool/main/analyse.cpp:15:
In file included from /tmp/tmp1i8ffxxt/src/src/buildtool/main/analyse.hpp:18:
In file included from /tmp/tmp1i8ffxxt/src/src/buildtool/build_engine/analysed_target/analysed_target.hpp:24:
In file included from /tmp/tmp1i8ffxxt/src/src/buildtool/build_engine/analysed_target/target_graph_information.hpp:21:
In file included from /tmp/tmp1i8ffxxt/src/src/buildtool/build_engine/target_map/configured_target.hpp:20:
/usr/include/fmt/core.h:2743:12: error: call to deleted constructor of 'conditional_t<has_formatter<mapped_type, context>::value, formatter<mapped_type, char_type>, fallback_formatter<stripped_type, char_type>>' (aka 'fmt::detail::fallback_formatter<LogLevel>')
  auto f = conditional_t<has_formatter<mapped_type, context>::value,
           ^
/usr/include/fmt/core.h:2954:23: note: in instantiation of function template specialization 'fmt::detail::parse_format_specs<LogLevel, fmt::detail::compile_parse_context<char>>' requested here
        parse_funcs_{&parse_format_specs<Args, parse_context_type>...},
                      ^
/usr/include/fmt/core.h:3159:47: note: in instantiation of member function 'fmt::detail::format_string_checker<char, fmt::detail::error_handler, LogLevel, LogLevel, LogLevel>::format_string_checker' requested here
      detail::parse_format_string<true>(str_, checker(s, {}));
                                              ^
/tmp/tmp1i8ffxxt/src/src/buildtool/common/cli.hpp:177:24: note: in instantiation of function template specialization 'fmt::basic_format_string<char, const LogLevel &, const LogLevel &, const LogLevel &>::basic_format_string<char[70], 0>' requested here
           fmt::format("Log limit (higher is more verbose) in interval [{},{}] "
                       ^
/usr/include/fmt/core.h:1124:3: note: 'fallback_formatter' has been explicitly marked deleted here
  fallback_formatter() = delete;
  ^
/usr/include/fmt/core.h:1756:3: error: static assertion failed due to requirement 'formattable': Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
  static_assert(
  ^
/usr/include/fmt/core.h:1777:10: note: in instantiation of function template specialization 'fmt::detail::make_value<fmt::basic_format_context<fmt::appender, char>, const LogLevel &>' requested here
  return make_value<Context>(val);
         ^
/usr/include/fmt/core.h:1899:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, fmt::detail::type::custom_type, const LogLevel &, 0>' requested here
        data_{detail::make_arg<
                      ^
/usr/include/fmt/core.h:1918:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, LogLevel, LogLevel, LogLevel>::format_arg_store<const LogLevel &, const LogLevel &, const LogLevel &>' requested here
  return {FMT_FORWARD(args)...};
         ^
/usr/include/fmt/core.h:3206:28: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::basic_format_context<fmt::appender, char>, const LogLevel &, const LogLevel &, const LogLevel &>' requested here
  return vformat(fmt, fmt::make_format_args(args...));
                           ^
/tmp/tmp1i8ffxxt/src/src/buildtool/common/cli.hpp:177:17: note: in instantiation of function template specialization 'fmt::format<const LogLevel &, const LogLevel &, const LogLevel &>' requested here
           fmt::format("Log limit (higher is more verbose) in interval [{},{}] "
                ^
2 errors generated.
```

As far as I understand, this points to a missing `fmt::formatter` implementation for `LogLevel`. So I added one, based on `LogLevelToString()` and the `fmt::formatter<std::string>`. This way, I was able to fix the issue. Here is the full code change: https://github.com/mhthies/justbuild/commit/548b35204dd62ac9df09d602ad2fdf6c51d21309

However, I'm not familiar with fmt, so I'm not sure if this is the correct fix. I also didn't test this change against fmt 7.x and other compilers.

**Comment by aehlig** [comment]
The offending call that would require a fallback formatter is the help text for the log level on the command line. There, however, we expect the numerical value. So, in my opinion, the correct solution is to [cast from the `enum class  LogLevel` back to int](https://github.com/just-buildsystem/justbuild/pull/15/commits/64f191efd26064a0ab85886ec3cb187ed7c2b738#diff-38adeb583131d399e9445e67ad732329013b2c4619cd73508ebb3feb651b3ec6).

In my opinion, #15 fixes this issue. What do you think?

**Comment by mhthies** [comment]
Ah, okay. This looks reasonable.

I didn't look up the code usage of the formatter, so I (wrongly) assumed the LogLevel was expected to be formatted as string. There would even have been a simpler solution for declaring the formatter. However, for formatting as numerical value, the explicit cast is definitely the better solution.

**Comment by aehlig** [comment]
As discussed, fixed by 4d5724b6d71f5251f8945056f278b3142f66f1b2
